### DISCUSSION TITLE
Hi @benolmbrt, thanks for your reporting.

### DISCUSSION THREAD

**Discussion by RY-97** [Project-MONAI/MONAI#8099] [General]
Hi @benolmbrt, thanks for your reporting.
The behavior you're seeing with the `DiceMetric` is likely due to a misunderstanding of the ignore_empty parameter's functionality.
When the `ignore_empty` option is set to True, it's supposed to set a NaN value for cases where the ground truth is empty. However, this does not mean that a NaN will be returned for such classes when computing the metric. Instead, these cases are ignored during the calculation process.Thus, even if some classes are ignored (due to being absent in the ground truth when ignore_empty=True), the metric computation does not return NaN for these classes. Instead, it computes the metric over the non-ignored classes only.
If the `DiceMetric` were to return a NaN for an ignored class, this would also affect the computed average Dice score (since any mathematical operation with NaN results in NaN), rendering the metric not well-defined for batches that include at least one empty class in the ground truth.

I hope this helps and let me know if you have any other questions!
Thanks.

_Originally posted by @KumoLiu in https://github.com/Project-MONAI/MONAI/issues/7504#issuecomment-1971455924_

**Comment by RY-97** [comment]
![image](https://github.com/user-attachments/assets/e3bf1fe8-b38a-4ec6-8804-b04c58090f68)
Why does this line of code produce "nan" when calculating, can you help me solve it?

**Comment by KumoLiu** [comment]
Hi @RY-97, please follow the example to generate Dice. Pay attention to the indentation.
```
        dice_metric = DiceMetric(include_background=True, reduction="mean")

        for val_data in val_loader:
            val_outputs = model(val_data["img"])
            val_outputs = [postprocessing_transform(i) for i in decollate_batch(val_outputs)]
            # compute metric for current iteration
            dice_metric(y_pred=val_outputs, y=val_data["seg"])  # callable to add metric to the buffer

        # aggregate the final mean dice result
        metric = dice_metric.aggregate().item()

        # reset the status for next computation round
        dice_metric.reset()
```
For more details you can refer to the tutorial repo:
https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/brats_segmentation_3d.ipynb

Hope it helps, thanks.

**Reply by RY-97** [reply]
![Uploading image.png…]()

**Reply by RY-97** [reply]
My input is correct, but nan appears when calculating DICE and HD

**Reply by KumoLiu** [reply]
https://github.com/Project-MONAI/MONAI/discussions/4540#discussioncomment-10700946

Based on the screenshot you shared, the reset operation is currently inside the loop. It should be moved outside of the loop.

**Reply by RY-97** [reply]
![image](https://github.com/user-attachments/assets/d364d359-af13-4e3e-b1a9-7b2d1952d820)
Is this correct?

**Reply by KumoLiu** [reply]
Also the aggregate, please check the example I shared. https://github.com/Project-MONAI/MONAI/discussions/8099#discussioncomment-10701198
Thanks.

**Reply by RY-97** [reply]
![image](https://github.com/user-attachments/assets/7e2ddb38-4ffd-4a9f-8401-7c3a2c81da3b)
If I change np.mean to np.nanmean, I can eliminate the nan situation in calculating certain categories of dcie and HD, but is this right?

**Reply by RY-97** [reply]
Because I have multiple categories, I need to get the dice and hd of each category, so the method you provided seems not supported. I tried it, thank you

**Reply by KumoLiu** [reply]
If you want the dice for each channel, please set the reduction to "mean_batch" instead of modify the indentation. Thanks.
https://github.com/Project-MONAI/MONAI/blob/d2d492ec045848b5872d8333e5174086c9cdfead/monai/metrics/meandice.py#L42

**Reply by RY-97** [reply]
Can you help me debug the following remotely?

**Reply by RY-97** [reply]
def validation(epoch_iterator_val, model, args):
    # 设置模型为验证模式
    model.eval()

    # 后处理函数，输出为 one-hot 编码
    post_label = AsDiscrete(to_onehot=args.num_classes)
    post_pred = AsDiscrete(argmax=True, to_onehot=args.num_classes)

    # 初始化度量指标
    dice_metric = DiceMetric(include_background=False, reduction="mean_batch", get_not_nans=False)
    hd_metric = HausdorffDistanceMetric(include_background=False, reduction="mean_batch", percentile=95)

    # 存储度量结果的列表
    dice_vals_per_class = []
    hd_vals_per_class = []

    with torch.no_grad():
        for step, batch in enumerate(epoch_iterator_val):
            val_inputs, val_labels = batch["image"].cuda(), batch["label"].cuda()

            # 使用滑动窗口推理
            val_outputs = sliding_window_inference(val_inputs, (96, 96, 96), 2, model)

            # 解码并转换标签和预测
            val_labels_list = decollate_batch(val_labels)
            val_labels_convert = [post_label(val_label_tensor) for val_label_tensor in val_labels_list]

            val_outputs_list = decollate_batch(val_outputs)
            val_outputs_convert = [post_pred(val_pred_tensor) for val_pred_tensor in val_outputs_list]

            #
            # for idx, pred in enumerate(val_labels_convert):
            #     if np.isnan(pred).any():
            #         print(f"Warning: Sample {idx} Predicted Output contains NaN values.")
            #     else:
            #         print(f"Sample {idx} Predicted Output is valid.")
            #
            #     # 可选：打印有效的像素数量
            #     active_pixels = np.sum(pred)
            #     print(f"Sample {idx} - Active Pixels Count: {active_pixels}")

            # 计算每个类别的Dice和HD
            dice_metric(y_pred=val_outputs_convert, y=val_labels_convert)
            hd_metric(y_pred=val_outputs_convert, y=val_labels_convert)

            # 获取每个类别的Dice和HD（如果返回的是元组，逐一处理每个张量）
            dice_per_class = [d.detach().cpu().numpy() for d in dice_metric.aggregate().item()]
            hd_per_class = [h.detach().cpu().numpy() for h in hd_metric.aggregate().item()]

            dice_vals_per_class.append(dice_per_class)
            hd_vals_per_class.append(hd_per_class)

            # 重置度量器
        dice_metric.reset()
        hd_metric.reset()

    # 转换为 NumPy 数组并计算平均值
    dice_vals_per_class = np.array(dice_vals_per_class)
    hd_vals_per_class = np.array(hd_vals_per_class)

    mean_dice_per_class = np.nanmean(dice_vals_per_class, axis=0)
    mean_hd_per_class = np.nanmean(hd_vals_per_class, axis=0)
    print(("--"*40) + "\n")
    # 输出每个类别的Dice和HD，确保格式化时使用标量
    for i in range(0, args.num_classes-1):
        print(f"Class {i} - Dice: {mean_dice_per_class[0][i].mean():.4f}, HD: {mean_hd_per_class[0][i].mean():.4f}")

    # 计算并返回整体的平均评价指标
    overall_dice = np.nanmean(mean_dice_per_class)
    overall_hd = np.nanmean(mean_hd_per_class)

    print(f"Overall Dice: {overall_dice:.4f}")
    print(f"Overall HD: {overall_hd:.4f}")

    # nib.save(nib.Nifti1Image(val_outputs_convert.astype(np.float32), np.eye(4)), "pred.nii.gz")

    return mean_dice_per_class, mean_hd_per_class, overall_dice, overall_hd

**Reply by RY-97** [reply]
This is my code, how should I modify it?

**Reply by RY-97** [reply]
can i help me ?
